sysvabi64/sysvabi64.rst

```diff
@@ -916,10 +923,639 @@ These follow the AAELF64_ base definition from Arm.
+
+A GNU Indirect Function (IFUNC) is a feature that permits a single
+implementation of a function to be chosen from multiple candidates,
+with the choice taken by an IFUNC resolver function.
+
+GNU Indirect Functions require static and dynamic linker support. They
+are known to be supported on GNU/Linux, Android and many of the BSD
+operating systems.
+
+GNU Indirect Functions are called via a PLT entry that loads the
+function address. The function address is chosen by an IFUNC resolver
+function.
+
+The source code interface to an IFUNC resolver is platform
+dependent. The GNU/Linux interface via the GNU C Library is documented
+below.
+
+GNU C Library IFUNC interface
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The GNU C Library (glibc) ``sys/ifunc.h`` header provides the
+necessary type definitions for defining IFUNC resolvers. The resolver
+is passed at least one, and at most 2 parameters dependent on the
+glibc IFUNC ABI version. A flag is set in the first parameter if a
+second parameter is available. The parameters provide access to the
+HWCAP_ flags.
+
+.. code-block:: c
+
+  /* A second argument is passed to the ifunc resolver.  */
+  #define _IFUNC_ARG_HWCAP        (1ULL << 62)
+
+  /* Function prototype for an IFUNC resolver.
+  ElfW(Addr) ifunc_resolver (uint64_t, const __ifunc_arg_t *);
+
+  /* Second argument to an ifunc resolver.  */
+  struct __ifunc_arg_t
+  {
+    unsigned long _size; /* Size of the struct, so it can grow.  */
+    unsigned long _hwcap;
+    unsigned long _hwcap2;
+  };
+
+  typedef struct __ifunc_arg_t __ifunc_arg_t;
+
+IFUNC resolver functions must have a type of ``STT_GNU_IFUNC``. With
+the GCC and Clang compilers an attribute can be used to achieve this.
+
+.. code-block:: c
+
+  /* Implementation alternative 1 */
+  static int implementation1(void) { ... }
+
+  /* Implementation alternative 2 */
+  static int implementation2(void) { ... }
+
+  /* Resolver function */
+  static void* resolver(uint64_t, const __ifunc_arg_t *) { ... }
+
+  /* make symbol ifunc type STT_GNU_IFUNC using resolver as the IFUNC resolver. */
+  int ifunc(void) __attribute__((ifunc("resolver")));
+
+The first ``uint64`` parameter is ``AT_HWCAP``. If the
+``_IFUNC_ARG_HWCAP`` flag is set the second parameter is present. The
+second parameter is a structure ``__ifunc_arg_t`` defined in
+``sys/ifunc.h`` it provides access to all of the flags in HWCAP_ via
+fields of the structure.
+
+The IFUNC resolver function returns the address of the function
+implementation.
+
+IFUNC resolvers may be run when the dynamic linker is resolving relocations.
+additional restrictions on what they can contain.
+
+ * IFUNC resolvers must not be compiled with security features like
+   stack-protection, which requires a guard variable to be
+   initialized. Or instrumentation like ASAN that requires a shadow
+   map to be set up.
+
+ * IFUNC resolvers must not have a symbol binding of ``STB_WEAK``.
+
+ * IFUNC resolvers that may be called via lazy-binding must be
```

Comment:
> IFUNC resolvers that may be called via lazy-binding must be asynchronous-signal safe and thread-safe as the resolver may be called from a signal handler or multi-threaded process.

The "async-signal safe and thread-safe" part on https://sourceware.org/glibc/wiki/GNU_IFUNC does not match the implementation. IFUNC resolvers are eagerly called even in the lazy binding PLT mode. Perhaps drop the description here.

Reply:
Thanks, I'll drop it. I'd thought that a `R_AARCH64_JUMP_SLOT` reloc would not be eagerly evaluated but it looks like, in glibc at least, that if there is a Relocation with respect to a symbol of type STT_GNU_IFUNC the resolver is called prior to processing the relocation.